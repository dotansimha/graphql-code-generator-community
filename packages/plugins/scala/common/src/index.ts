export * from './config';
export * from './base-visitor';
export * from './base-plugin';
