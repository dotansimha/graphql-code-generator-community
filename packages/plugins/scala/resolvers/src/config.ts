import { ScalaPluginCommonRawConfig } from '@graphql-codegen/scala-common';

/**
 * @description This plugin generates Scala Sangria resolver files.
 */
export interface ScalaResolversPluginRawConfig extends ScalaPluginCommonRawConfig {
  /**
   * @description The package name to use in the generated files.
   * @default - Derived from the output file path
   */
  packageName?: string;

  /**
   * @description The class name to use for the main class.
   * @default 'Types'
   */
  className?: string;

  /**
   * @description Use Scala Future types as wrapper for resolver results
   * @default false
   */
  withFuture?: boolean;

  /**
   * @description Use Scala ZIO types as wrapper for resolver results
   * @default false
   */
  withZIO?: boolean;

  /**
   * @description Prefix for class member names
   * @default ''
   */
  classMembersPrefix?: string;

  /**
   * @description Type to use for GraphQL lists
   * @default 'List'
   */
  listType?: string;

  /**
   * @description Use Option for nullable fields
   * @default false
   */
  useOptions?: boolean;

  /**
   * @description Package name for model classes
   * @default - Same as packageName
   */
  modelPackage?: string;
}

export interface ScalaResolversPluginParsedConfig {
  className: string;
  withFuture: boolean;
  withZIO: boolean;
  classMembersPrefix: string;
  listType: string;
  useOptions: boolean;
  package?: string;  // To support legacy code
  modelPackage?: string;
}
