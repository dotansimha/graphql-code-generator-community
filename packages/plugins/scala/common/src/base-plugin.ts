import { GraphQLSchema } from 'graphql';
import { Types } from '@graphql-codegen/plugin-helpers';
import { ScalaBaseVisitor } from './base-visitor';
import { ScalaPluginCommonRawConfig } from './config';

export type BasePluginFunction<TRawConfig extends ScalaPluginCommonRawConfig> = (
  schema: GraphQLSchema,
  documents: Types.DocumentFile[],
  config: TRawConfig,
  info: { outputFile?: string },
) => Types.PluginOutput;

export const createBasePlugin = <
  TRawConfig extends ScalaPluginCommonRawConfig,
  TVisitor extends ScalaBaseVisitor<TRawConfig, any>,
>(
  createVisitor: (
    schema: GraphQLSchema,
    config: TRawConfig,
    outputFileInfo: { outputFile?: string },
  ) => TVisitor,
): BasePluginFunction<TRawConfig> => {
  return (
    schema: GraphQLSchema,
    _: Types.DocumentFile[],
    config: TRawConfig,
    info: { outputFile?: string },
  ): Types.PluginOutput => {
    if (!config.packageName && info.outputFile) {
      const parts = info.outputFile.split('/');
      parts.pop();

      if (parts.length > 0) {
        config = {
          ...config,
          packageName: parts.join('.'),
        };
      }
    }

    const visitor = createVisitor(schema, config, { outputFile: info.outputFile });
    const typeNames = Object.keys(schema.getTypeMap()).filter(
      typeName => !typeName.startsWith('__'),
    );

    const header = [
      visitor.getPackage(),
      '',
      visitor.getImports(),
      '',
      'getTypeClassDefinitions' in visitor ? (visitor as any).getTypeClassDefinitions() : '',
    ]
      .filter(Boolean)
      .join('\n');

    const schemaScalarTypeNames = typeNames.filter(typeName => {
      const type = schema.getTypeMap()[typeName];
      return type.astNode?.kind === 'ScalarTypeDefinition';
    });

    const customScalarDefinitions: string[] = [];
    if (config.scalars) {
      Object.keys(config.scalars).forEach(scalarName => {
        if (typeof config.scalars[scalarName] === 'string') {
          const scalarType = config.scalars[scalarName];
          customScalarDefinitions.push(`type ${scalarName} = ${scalarType}`);
        }
      });
    }

    const defaultScalars = visitor.generateCustomScalars
      ? visitor.generateCustomScalars([
          ...schemaScalarTypeNames,
          ...Object.keys(config.scalars || {}),
        ])
      : [];

    const contentFragments = typeNames.map(typeName => {
      const type = schema.getTypeMap()[typeName];

      if (
        type.astNode?.kind === 'ScalarTypeDefinition' &&
        config.scalars &&
        typeof config.scalars[typeName] === 'string'
      ) {
        return '';
      }

      const astNodeKind = type.astNode?.kind as string | undefined;
      if (astNodeKind && astNodeKind in visitor) {
        const handler = (visitor as any)[astNodeKind];
        if (typeof handler === 'function') {
          return handler.call(visitor, type.astNode);
        }
      }

      return '';
    });

    const content = [
      header,
      ...customScalarDefinitions,
      ...defaultScalars,
      ...contentFragments.filter(Boolean),
    ].join('\n\n');

    return {
      content,
      prepend: [],
      append: [],
    };
  };
};
