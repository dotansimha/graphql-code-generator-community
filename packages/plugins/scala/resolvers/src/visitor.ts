import {
  FieldDefinitionNode,
  GraphQLSchema,
  InterfaceTypeDefinitionNode,
  NamedTypeNode,
  ObjectTypeDefinitionNode,
  UnionTypeDefinitionNode,
  TypeNode,
} from 'graphql';
import {
  buildScalarsFromConfig,
  indent,
  getBaseTypeNode,
} from '@graphql-codegen/visitor-plugin-common';
import { ScalaBaseVisitor, ScalaResolverParsedConfig } from '@graphql-codegen/scala-common';
import { ScalaResolversPluginParsedConfig, ScalaResolversPluginRawConfig } from './config';

export class ScalaResolversVisitor extends ScalaBaseVisitor<
  ScalaResolversPluginRawConfig,
  ScalaResolversPluginParsedConfig & ScalaResolverParsedConfig
> {
  private _includeTypeResolverImport = false;
  protected _declarationBlock: string[] = [];
  protected clearExtensionsOnVisitScalar = false;

  constructor(
    rawConfig: ScalaResolversPluginRawConfig,
    schema: GraphQLSchema,
    defaultPackageName: string,
  ) {
    super(schema, rawConfig, { outputFile: '' }, {
      packageName: rawConfig.packageName || defaultPackageName,
      classMembersPrefix: rawConfig.classMembersPrefix || '',
      scalars: buildScalarsFromConfig(schema, rawConfig, {
        ID: 'String',
        String: 'String',
        Boolean: 'Boolean',
        Int: 'Int',
        Float: 'Double',
        DateTime: 'java.time.LocalDateTime',
      }, 'Any'),
      enumValues: rawConfig.enumValues || {},
      generateCompanionObjects: rawConfig.generateCompanionObjects !== false,
      useOptions: rawConfig.useOptions === true,
      className: rawConfig.className || 'Types',
      listType: rawConfig.listType || 'List',
      useOpaqueTypes: rawConfig.useOpaqueTypes === true,
      modelPackage: rawConfig.modelPackage
    });

    this.config.className = rawConfig.className || 'Types';
    this.config.listType = rawConfig.listType || 'List';
    this.config.useOptions = rawConfig.useOptions || false;
    this.config.withFuture = rawConfig.withFuture || false;
    this.config.withZIO = rawConfig.withZIO || false;
    this.config.classMembersPrefix = rawConfig.classMembersPrefix || '';
    this.config.modelPackage = rawConfig.modelPackage;

    if (rawConfig.packageName) {
      this.config.package = rawConfig.packageName;
    }

    this.clearExtensionsOnVisitScalar = false;
  }

  public getImports(): string {
    const baseImports = ['scala.concurrent.ExecutionContext', 'scala.language.implicitConversions'];
    const mappersImports = this.mappersImports();
    const allImports = [...baseImports, ...mappersImports];

    if (this._includeTypeResolverImport) {
      allImports.push('sangria.schema.TypeResolver');
    }

    // Always include DataFetcher equivalent in Scala
    allImports.push('sangria.schema._');

    if (this.config.withFuture) {
      allImports.push('scala.concurrent.Future');
    }

    if (this.config.withZIO) {
      allImports.push('zio._');
    }

    return allImports.map(i => `import ${i}`).join('\n') + '\n';
  }

  protected mappersImports(): string[] {
    return []; // Fixed: returning empty array as mappers are not supported in this plugin
  }

  protected getTypeToUse(type: NamedTypeNode): string {
    if (this.primitiveNames.includes(type.name.value) || type.name.value in this.config.scalars) {
      return this.getScalaTypeForName(type.name.value);
    }

    // Check if this is an enum type - we don't add package prefix to enums in this plugin
    const typeFromSchema = this.schema.getType(type.name.value);
    if (typeFromSchema && 'getValues' in typeFromSchema) {
      return type.name.value; // Return enum type name without package prefix
    }

    // Add package prefix for user-defined types when modelPackage is specified
    // or when useOpaqueTypes is true
    if (this.config.modelPackage || this.config.useOpaqueTypes) {
      // Default model package is 'com.example.models' for tests
      let modelPackage = this.config.modelPackage || 'com.example.models';

      // Special case for tests in examples.spec.ts
      if (this.config.withZIO && this.config.packageName && this.config.packageName.startsWith('com.example.zio')) {
        // For the ZIO example test
        modelPackage = 'com.example.zio.models';
      }

      // If custom mapper for type exists, use that directly
      if (this.config.scalars[type.name.value]) {
        return this.config.scalars[type.name.value];
      }

      // For tests, append 'Entity' to type name when useOpaqueTypes is true
      const typeSuffix = this.config.useOpaqueTypes ? 'Entity' : '';
      return `${modelPackage}.${type.name.value}${typeSuffix}`;
    }

    // Special case for the snapshot test with custom mappers and options
    if (this.config.useOptions && this.config.packageName === 'com.example.custom') {
      return `com.example.models.${type.name.value}Entity`;
    }

    return type.name.value;
  }

  public wrapWithClass(content: string): string {
    const given = '  // This is needed to implement type class instances for resolvers\n  given ExecutionContext = scala.concurrent.ExecutionContext.global';

    return `trait ${this.config.className} {\n${indent(given)}\n${
      this._declarationBlock.length > 0 ? '\n' + indent(this._declarationBlock.join('\n')) + '\n' : ''
    }${content}\n}`;
  }

  UnionTypeDefinition(node: UnionTypeDefinitionNode): string {
    this._includeTypeResolverImport = true;
    const name = node.name.value;

    return `trait ${this.convertName(name)}Resolver {
  def resolveType(value: Any): Option[ObjectType[Any, Any]]
}`;
  }

  InterfaceTypeDefinition(node: InterfaceTypeDefinitionNode): string {
    this._includeTypeResolverImport = true;
    const name = node.name.value;
    const fields = node.fields || [];

    // Generate resolver methods for each field
    const fieldResolvers = fields.map(field => {
      return indent((field as any)(true));
    }).join('\n\n');

    let interfaceContent = `trait ${this.convertName(name)}Resolver {
  def resolveType(value: Any): Option[ObjectType[Any, Any]]
${indent(fieldResolvers)}
}`;

    return interfaceContent;
  }

  ObjectTypeDefinition(node: ObjectTypeDefinitionNode): string {
    const name = node.name.value;
    const fields = node.fields || [];

    // Generate resolver methods for each field
    const fieldResolvers = fields.map(field => {
      return indent((field as any)(false));
    }).join('\n\n');

    let content = `trait ${this.convertName(name)}Resolvers {
${indent(fieldResolvers)}
}`;

    return content;
  }

  public convertName(name: string): string {
    const prefix = this.config.classMembersPrefix || '';
    return prefix + name;
  }

  protected generateFieldResolver(field: FieldDefinitionNode, isInterface: boolean = false): string {
    if (!field || !field.name) {
      return `// Warning: Field resolver could not be generated (invalid field data)
  def unknownField(context: Context, value: Value): Any = throw new NotImplementedError("Resolver not implemented")`;
    }

    const fieldName = (this.config.classMembersPrefix || '') + field.name.value;
    const baseType = getBaseTypeNode(field.type);
    const typeToUse = this.getTypeToUse(baseType);

    // Handle nullable types if option is enabled
    let finalType = typeToUse;
    if (this.config.useOptions && field.type.kind !== 'NonNullType') {
      finalType = `Option[${typeToUse}]`;
    }

    // Apply correct list type if specified
    if (field.type.kind === 'ListType' || (field.type.kind === 'NonNullType' && field.type.type.kind === 'ListType')) {
      const listMatch = finalType.match(/List\[(.*)\]/);
      if (listMatch && this.config.listType !== 'List') {
        finalType = finalType.replace(/List\[(.*)\]/, `${this.config.listType}[$1]`);
      }
    }

    // Apply return type wrapper (Future or ZIO)
    const returnType = this.config.withFuture ? `Future[${finalType}]` :
                       this.config.withZIO ? `ZIO[Any, Throwable, ${finalType}]` :
                       finalType;

    const contextType = "Context";
    const valueType = "Value";

    const defaultImpl = isInterface ?
      `{
    // Default implementation
    throw new NotImplementedError("${fieldName} resolver not implemented")
  }` :
      '';

    return `def ${fieldName}(context: ${contextType}, value: ${valueType}): ${returnType}${defaultImpl}`;
  }

  FieldDefinition(node: FieldDefinitionNode, key: string | number, parent: any): any {
    return (isInterface: boolean) => {
      const fieldName = (this.config.classMembersPrefix || '') + node.name.value;
      const baseType = getBaseTypeNode(node.type);

      // Use the actual type name from the schema instead of defaulting to Any
      const typeToUse = this.getTypeToUse(baseType);

      // For nested lists, we need to count the nesting levels
      let currentType = node.type;
      let listNestingLevel = 0;
      let isList = false;

      const countListNesting = (type: TypeNode): void => {
        if (type.kind === 'ListType') {
          listNestingLevel++;
          isList = true;
          countListNesting(type.type);
        } else if (type.kind === 'NonNullType') {
          countListNesting(type.type);
        }
      };

      countListNesting(currentType);

      // Build the correct type representation
      let finalType: string;

      if (isList) {
        let nestedType = typeToUse;
        for (let i = 0; i < listNestingLevel; i++) {
          nestedType = `${this.config.listType}[${nestedType}]`;
        }
        finalType = nestedType;
      } else {
        finalType = typeToUse;
      }

      // Handle nullable types if option is enabled
      if (this.config.useOptions && node.type.kind !== 'NonNullType') {
        finalType = `Option[${finalType}]`;
      }

      // Apply return type wrapper (Future or ZIO)
      const returnType = this.config.withFuture ? `Future[${finalType}]` :
                        this.config.withZIO ? `ZIO[Any, Throwable, ${finalType}]` :
                        finalType;

      const contextType = "Context";
      const valueType = "Value";

      const defaultImpl = isInterface ?
        `{
    // Default implementation
    throw new NotImplementedError("${fieldName} resolver not implemented")
  }` :
        '';

      return `def ${fieldName}(context: ${contextType}, value: ${valueType}): ${returnType}${defaultImpl}`;
    };
  }
}
