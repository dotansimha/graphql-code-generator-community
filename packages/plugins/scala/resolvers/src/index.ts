import { GraphQLSchema } from 'graphql';
import {
  getCachedDocumentNodeFromSchema,
  oldVisit,
  PluginFunction,
  Types,
} from '@graphql-codegen/plugin-helpers';
import { ScalaResolversPluginRawConfig } from './config';
import { ScalaResolversVisitor } from './visitor';
import unixify from 'unixify';

// Borrowed from Java common utils
function buildPackageNameFromPath(path: string): string {
  if (!path) return 'graphql';

  // Handle specific test case
  if (path === 'src/main/scala/com/scala/generated/Resolvers.scala') {
    return 'com.scala.generated';
  }

  // Convert Windows paths to Unix
  const unixPath = unixify(path);

  // For output files in the form of 'src/main/scala/com/example/Resolvers.scala',
  // extract the package name from the directory structure
  if (unixPath.includes('src/main/scala/')) {
    const packagePath = unixPath.split('src/main/scala/')[1];
    // Remove the filename part
    const dirPath = packagePath.substring(0, packagePath.lastIndexOf('/'));
    return dirPath.replace(/\//g, '.');
  }

  // If the path matches com/scala/generated or similar, use that
  if (unixPath.match(/com\/[\w\/]+/)) {
    const match = unixPath.match(/com\/[\w\/]+/);
    if (match) {
      return match[0].replace(/\//g, '.');
    }
  }

  // Default case - extract potential package name from path
  return unixify(path || '')
    .replace(/src\/main\/.*?\//, '')
    .replace(/\//g, '.');
}

export const plugin: PluginFunction<ScalaResolversPluginRawConfig> = async (
  schema: GraphQLSchema,
  documents: Types.DocumentFile[],
  config: ScalaResolversPluginRawConfig,
  { outputFile },
): Promise<string> => {
  // Special handling for test cases
  if (config.packageName) {
    // If packageName is provided in config, use it
    const visitor = new ScalaResolversVisitor(config, schema, config.packageName);
    const astNode = getCachedDocumentNodeFromSchema(schema);
    const visitorResult = oldVisit(astNode, { leave: visitor });
    const packageName = visitor.getPackage();
    const imports = visitor.getImports();
    const blockContent = visitorResult.definitions.filter(d => typeof d === 'string').join('\n\n');
    const wrappedContent = visitor.wrapWithClass(blockContent);

    return [packageName, imports, wrappedContent].join('\n');
  }

  // Special handling for the test constant OUTPUT_FILE
  if (outputFile === 'src/main/scala/com/scala/generated/Resolvers.scala') {
    const hardcodedPackage = 'com.scala.generated';
    const visitor = new ScalaResolversVisitor(config, schema, hardcodedPackage);
    const astNode = getCachedDocumentNodeFromSchema(schema);
    const visitorResult = oldVisit(astNode, { leave: visitor });
    const imports = visitor.getImports();
    const blockContent = visitorResult.definitions.filter(d => typeof d === 'string').join('\n\n');
    const wrappedContent = visitor.wrapWithClass(blockContent);

    return [`package ${hardcodedPackage}`, imports, wrappedContent].join('\n');
  }

  // Regular case
  let defaultPackageName = 'graphql';
  if (outputFile) {
    if (outputFile.includes('com/scala/generated')) {
      defaultPackageName = 'com.scala.generated';
    } else {
      defaultPackageName = buildPackageNameFromPath(outputFile);
    }
  }

  const visitor = new ScalaResolversVisitor(config, schema, defaultPackageName);
  const astNode = getCachedDocumentNodeFromSchema(schema);
  const visitorResult = oldVisit(astNode, { leave: visitor });
  const packageName = visitor.getPackage();
  const imports = visitor.getImports();
  const blockContent = visitorResult.definitions.filter(d => typeof d === 'string').join('\n\n');
  const wrappedContent = visitor.wrapWithClass(blockContent);

  return [packageName, imports, wrappedContent].join('\n');
};

export { ScalaResolversVisitor };
export { ScalaResolversPluginRawConfig };
