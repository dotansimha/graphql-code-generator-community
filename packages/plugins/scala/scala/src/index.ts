import { PluginFunction } from '@graphql-codegen/plugin-helpers';
import { createBasePlugin } from '@graphql-codegen/scala-common';
import { ScalaPluginRawConfig } from './config';
import { ScalaVisitor } from './visitor';

export const plugin: PluginFunction<ScalaPluginRawConfig> = createBasePlugin<
  ScalaPluginRawConfig,
  ScalaVisitor
>((schema, config, info) => new ScalaVisitor(schema, config, info));

export { ScalaPluginRawConfig } from './config';
