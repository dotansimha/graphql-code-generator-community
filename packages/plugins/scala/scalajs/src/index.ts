import { PluginFunction } from '@graphql-codegen/plugin-helpers';
import { createBasePlugin } from '@graphql-codegen/scala-common';
import { ScalaJsPluginRawConfig } from './config';
import { ScalaJsVisitor } from './visitor';

export const plugin: PluginFunction<ScalaJsPluginRawConfig> = createBasePlugin<
  ScalaJsPluginRawConfig,
  ScalaJsVisitor
>((schema, config, info) => new ScalaJsVisitor(schema, config, info));

export { ScalaJsPluginRawConfig } from './config';
