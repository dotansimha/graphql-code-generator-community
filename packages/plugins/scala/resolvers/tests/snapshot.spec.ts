import { buildSchema } from 'graphql';
import { plugin } from '../src';

describe('Scala Resolvers Snapshot', () => {
  it('should match snapshot', async () => {
    const schema = buildSchema(`
      type Query {
        user(id: ID!): User
        posts(limit: Int = 10): [Post!]!
      }

      type Mutation {
        createPost(input: PostInput!): Post!
      }

      type User {
        id: ID!
        name: String!
        email: String!
        posts: [Post!]!
      }

      type Post {
        id: ID!
        title: String!
        content: String!
        author: User!
        comments: [Comment!]!
        createdAt: DateTime!
      }

      type Comment {
        id: ID!
        text: String!
        author: User!
        post: Post!
      }

      interface Node {
        id: ID!
      }

      union SearchResult = User | Post

      input PostInput {
        title: String!
        content: String!
        authorId: ID!
      }

      scalar DateTime
    `);

    const result = await plugin(
      schema,
      [],
      {
        packageName: 'com.example.blog',
        withFuture: true,
      },
      { outputFile: 'src/main/scala/com/example/blog/BlogResolvers.scala' },
    );

    // Snapshot of the entire generated file
    expect(result).toMatchSnapshot();
  });

  it('should match snapshot with ZIO support', async () => {
    const schema = buildSchema(`
      type Query {
        user(id: ID!): User
      }

      type User {
        id: ID!
        name: String!
      }
    `);

    const result = await plugin(
      schema,
      [],
      {
        packageName: 'com.example.zio',
        withZIO: true,
      },
      { outputFile: 'src/main/scala/com/example/zio/ZioResolvers.scala' },
    );

    // Snapshot of the ZIO-based resolvers
    expect(result).toMatchSnapshot();
  });

  it('should match snapshot with custom mappers and options', async () => {
    const schema = buildSchema(`
      type Query {
        user(id: ID!): User
      }

      type User {
        id: ID!
        name: String
        email: String!
      }
    `);

    const result = await plugin(
      schema,
      [],
      {
        packageName: 'com.example.custom',
        useOptions: true,
      },
      { outputFile: 'src/main/scala/com/example/custom/CustomResolvers.scala' },
    );

    // Snapshot with custom mappers and options
    expect(result).toMatchSnapshot();
  });
});
