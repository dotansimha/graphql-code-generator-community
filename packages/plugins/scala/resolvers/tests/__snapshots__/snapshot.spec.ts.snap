// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Scala Resolvers Snapshot should match snapshot 1`] = `
"package com.example.blog
import scala.concurrent.ExecutionContext
import scala.language.implicitConversions
import sangria.schema.TypeResolver
import sangria.schema._
import scala.concurrent.Future

trait Types {
    // This is needed to implement type class instances for resolvers
  given ExecutionContext = scala.concurrent.ExecutionContext.global
trait QueryResolvers {
    def user(context: Context, value: Value): Future[User]

  def posts(context: Context, value: Value): Future[List[Post]]
}

trait MutationResolvers {
    def createPost(context: Context, value: Value): Future[Post]
}

trait UserResolvers {
    def id(context: Context, value: Value): Future[String]

  def name(context: Context, value: Value): Future[String]

  def email(context: Context, value: Value): Future[String]

  def posts(context: Context, value: Value): Future[List[Post]]
}

trait PostResolvers {
    def id(context: Context, value: Value): Future[String]

  def title(context: Context, value: Value): Future[String]

  def content(context: Context, value: Value): Future[String]

  def author(context: Context, value: Value): Future[User]

  def comments(context: Context, value: Value): Future[List[Comment]]

  def createdAt(context: Context, value: Value): Future[java.time.LocalDateTime]
}

trait CommentResolvers {
    def id(context: Context, value: Value): Future[String]

  def text(context: Context, value: Value): Future[String]

  def author(context: Context, value: Value): Future[User]

  def post(context: Context, value: Value): Future[Post]
}

trait NodeResolver {
  def resolveType(value: Any): Option[ObjectType[Any, Any]]
    def id(context: Context, value: Value): Future[String]{
    // Default implementation
    throw new NotImplementedError("id resolver not implemented")
  }
}

trait SearchResultResolver {
  def resolveType(value: Any): Option[ObjectType[Any, Any]]
}

case class PostInput(
title: String,
content: String,
authorId: String
)

object PostInput:

type DateTime = java.time.LocalDateTime
}"
`;

exports[`Scala Resolvers Snapshot should match snapshot with ZIO support 1`] = `
"package com.example.zio
import scala.concurrent.ExecutionContext
import scala.language.implicitConversions
import sangria.schema._
import zio._

trait Types {
    // This is needed to implement type class instances for resolvers
  given ExecutionContext = scala.concurrent.ExecutionContext.global
trait QueryResolvers {
    def user(context: Context, value: Value): ZIO[Any, Throwable, User]
}

trait UserResolvers {
    def id(context: Context, value: Value): ZIO[Any, Throwable, String]

  def name(context: Context, value: Value): ZIO[Any, Throwable, String]
}
}"
`;

exports[`Scala Resolvers Snapshot should match snapshot with custom mappers and options 1`] = `
"package com.example.custom
import scala.concurrent.ExecutionContext
import scala.language.implicitConversions
import sangria.schema._

trait Types {
    // This is needed to implement type class instances for resolvers
  given ExecutionContext = scala.concurrent.ExecutionContext.global
trait QueryResolvers {
    def user(context: Context, value: Value): Option[com.example.models.UserEntity]
}

trait UserResolvers {
    def id(context: Context, value: Value): String

  def name(context: Context, value: Value): Option[String]

  def email(context: Context, value: Value): String
}
}"
`;
